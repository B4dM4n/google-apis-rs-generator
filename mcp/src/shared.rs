//! This module, in some way or form, should contain all logic used to generate names.
//! These must be reused throughout the library.
use failure::Error;

pub fn crate_name(name: &str, version: &str) -> Result<String, Error> {
    make_target(name, version).map(|n| format!("google-{}", n))
}

pub fn make_target(name: &str, version: &str) -> Result<String, Error> {
    unimplemented!()
}

pub fn parse_version(version: &str) -> Result<String, Error> {
    Ok(version.into())
}

#[cfg(test)]
mod tests {
    mod crate_name {
        use crate::shared::crate_name;

        #[test]
        fn it_produces_a_valid_crate_name() {
            assert_eq!(crate_name("youtube", "v1.3").unwrap(), "google-youtube1")
        }
    }
    mod make_target {
        use crate::shared::make_target;

        #[test]
        fn it_produces_a_valid_make_target() {
            assert_eq!(make_target("youtube", "v1.3").unwrap(), "youtube1")
        }
    }
    mod parse_version {
        use super::super::parse_version;
        use insta::assert_snapshot_matches;
        use itertools::Itertools;
        use std::io::{BufRead, BufReader};

        const KNOWN_VERSIONS: &str = include_str!("../../tests/mcp/fixtures/shared/known-versions");

        #[test]
        fn it_works_for_all_known_inputs() {
            let expected = BufReader::new(KNOWN_VERSIONS.as_bytes())
                .lines()
                .filter_map(Result::ok)
                .map(|api_version| {
                    format!(
                        "{input} {output}",
                        input = api_version,
                        output = parse_version(&api_version).unwrap()
                    )
                })
                .join("\n");
            assert_snapshot_matches!(expected);
        }
    }
}
