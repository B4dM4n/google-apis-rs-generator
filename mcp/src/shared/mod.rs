//! This module, in some way or form, should contain all logic used to generate names.
//! These must be reused throughout the library.
use discovery_parser::generated::{ApiIndexV1, Item};
use failure::{bail, format_err, Error};
use serde::{Deserialize, Serialize};
use std::convert::TryFrom;
use std::path::PathBuf;

#[derive(Serialize, Deserialize)]
pub struct MappedIndex {
    pub api: Vec<Api>,
}

#[derive(Serialize, Deserialize)]
pub struct Api {
    pub name: String,
    pub gen_dir: PathBuf,
    pub spec_file: PathBuf,
    pub crate_name: String,
    pub make_target: String,
    pub rest_url: String,
}

impl TryFrom<Item> for Api {
    type Error = Error;

    fn try_from(value: Item) -> Result<Self, Self::Error> {
        let name = sanitized_name(&value.name).into();
        let gen_dir = PathBuf::from(&name).join(&value.version);
        Ok(Api {
            spec_file: gen_dir.join("spec.json"),
            gen_dir,
            name,
            rest_url: value.discovery_rest_url,
            crate_name: crate_name(&value.name, &value.version)?,
            make_target: make_target(&value.name, &value.version)?,
        })
    }
}

impl MappedIndex {
    pub fn from_index(index: ApiIndexV1) -> Result<Self, Error> {
        Ok(MappedIndex {
            api: index
                .items
                .into_iter()
                .map(Api::try_from)
                .collect::<Result<Vec<_>, Error>>()?,
        })
    }
}

pub fn crate_name(name: &str, version: &str) -> Result<String, Error> {
    make_target(name, version).map(|n| format!("google-{}", n))
}

pub fn sanitized_name(name: &str) -> &str {
    if let Some(pos) = name.rfind(|c| !char::is_digit(c, 10)) {
        &name[..=pos]
    } else {
        name
    }
}

pub fn make_target(name: &str, version: &str) -> Result<String, Error> {
    Ok(format!(
        "{name}{version}",
        name = sanitized_name(name),
        version = parse_version(version)?
    ))
}

pub fn parse_version(version: &str) -> Result<String, Error> {
    let inner = |version: &str| {
        if version.len() < 2 {
            bail!("version string too small");
        }
        if !version.is_ascii() {
            bail!("can only handle ascii versions");
        }
        if version == "alpha" || version == "beta" {
            return Ok(version.into());
        }

        fn transform_version(version: &str) -> Result<String, Error> {
            let mut bytes = version.bytes();
            if bytes.next() != Some(b'v') {
                bail!("A version must start with 'v'");
            }
            let mut out = String::new();
            let mut separator = Some('_');
            for b in bytes {
                let c = match b {
                    b'.' => b'd',
                    b @ b'0'..=b'9' => b,
                    b @ b'a'..=b'z' => {
                        if let Some(sep) = separator.take() {
                            out.push(sep);
                        }
                        b
                    }
                    b => bail!("unexpected character '{}'", b),
                } as char;
                out.push(c);
            }
            Ok(out)
        }

        let mut tokens = version.splitn(2, '_');
        if let (Some(left), Some(right)) = (tokens.next(), tokens.next()) {
            return Ok(format!(
                "{version}_{name}",
                version = transform_version(right)?,
                name = left
            ));
        }
        transform_version(version)
    };
    inner(version).map_err(|e| format_err!("invalid version '{}': {}", version, e))
}

#[cfg(test)]
mod tests;
