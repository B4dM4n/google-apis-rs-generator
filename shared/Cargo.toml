[package]
name = "shared"
version = "0.1.0"
authors = ["Sebastian Thiel <byronimo@gmail.com>"]
edition = "2018"
publish = false

[lib]
doctest = false

[dependencies]
serde = { version = "1", features = ["derive"] }
discovery_parser = { path = "../discovery_parser" }
failure = "0.1.5"
log = "0.4.8"
ci_info = "0.8.0"

[dev-dependencies]
itertools = "0.8.0"
insta = "0.10.1"

