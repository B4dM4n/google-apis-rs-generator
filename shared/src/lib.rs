//! This module, in some way or form, should contain all logic used to generate names.
//! These must be reused throughout the library.
//! You will find all/most of the constants here.
use discovery_parser::{
    generated::{ApiIndexV1, Icons, Item, Kind},
    DiscoveryRestDesc,
};
use failure::{bail, format_err, Error};
use log::{error, info};
use serde::{Deserialize, Serialize};
use std::{convert::TryFrom, path::Path, path::PathBuf};

/// A bunch of constants which must be the single source for constants
/// that are not API specific.
#[derive(Serialize, Deserialize)]
pub struct Standard {
    /// A cargo project relative path to the manifest file
    pub cargo_toml_path: String,
    /// A project relative path to the Rust library implementation
    pub lib_path: String,
    /// The name of the folder into which we want to generate the library projects
    pub lib_dir: String,
    /// The name of the folder containing specification files, as seen from the 'generated' repository
    pub spec_dir: String,
}

impl Default for Standard {
    fn default() -> Self {
        Standard {
            cargo_toml_path: "Cargo.toml".into(),
            lib_path: "src/lib.rs".into(),
            lib_dir: "lib".into(),
            spec_dir: "etc/api".into(),
        }
    }
}

#[derive(Serialize, Deserialize)]
pub struct MappedIndex {
    pub standard: Standard,
    pub api: Vec<Api>,
}

#[derive(Serialize, Deserialize)]
pub struct Api {
    /// The sanitized API name. See `sanitized_name(...)` for more information
    pub name: String,
    /// The official API id, good for identification
    pub id: String,
    /// A 'gen' directory relative path to the project manifest
    pub lib_cargo_file: PathBuf,
    /// A 'gen' directory relative path to the file containing any kind of generation or build error
    pub error_file: PathBuf,
    /// A 'gen' directory relative path into which all files pertaining the API must be placed
    pub gen_dir: PathBuf,
    /// A 'gen' directory relative path to the google discovery specification file
    pub spec_file: PathBuf,
    /// A suitable name for the crate
    pub crate_name: String,
    /// A suitable name for being a target in 'make'
    pub make_target: String,
    /// The URL to the google discovery specification
    pub rest_url: String,
}

impl TryFrom<Item> for Api {
    type Error = Error;

    fn try_from(value: Item) -> Result<Self, Self::Error> {
        let name = sanitized_name(&value.name).into();
        let gen_dir = PathBuf::from(&name).join(&value.version);
        let standard = Standard::default();
        Ok(Api {
            spec_file: gen_dir.join("spec.json"),
            id: value.id,
            lib_cargo_file: gen_dir
                .join(standard.lib_dir)
                .join(standard.cargo_toml_path),
            error_file: gen_dir.join("generator-errors.log"),
            gen_dir,
            name,
            rest_url: value.discovery_rest_url,
            crate_name: crate_name(&value.name, &value.version)?,
            make_target: make_target(&value.name, &value.version)?,
        })
    }
}

impl TryFrom<&DiscoveryRestDesc> for Api {
    type Error = Error;

    fn try_from(value: &DiscoveryRestDesc) -> Result<Self, Self::Error> {
        let item = Item {
            kind: Kind::DiscoveryDirectoryItem,
            id: value.id.to_owned(),
            name: value.name.to_owned(),
            version: value.version.to_owned(),
            title: value.title.to_owned(),
            description: value.description.to_owned(),
            discovery_rest_url: "<unset>".into(),
            icons: Icons {
                x16: "".to_string(),
                x32: "".to_string(),
            },
            documentation_link: None,
            preferred: false,
            discovery_link: None,
            labels: None,
        };
        Api::try_from(item)
    }
}

impl MappedIndex {
    pub fn validated(mut self, output_directory: &Path) -> Self {
        let ci_whitelist = ["admin:directory_v1", "compute:v1", "drive:v3", "oauth2:v2"];
        let info = ci_info::get();
        if info.ci {
            info!(
                "Running on CI '{:?}' - limiting APIs to {whitelist:?}",
                info.vendor,
                whitelist = ci_whitelist
            );
        }
        self.api.retain(|api| {
            let spec_path = output_directory.join(&api.spec_file);
            let is_allowed = if info.ci {
                ci_whitelist.contains(&api.id.as_str())
            } else {
                true
            };
            if !is_allowed {
                return false;
            }
            if !spec_path.is_file() {
                error!(
                    "Dropping API '{}' as its spec file at '{}' does not exist",
                    api.crate_name,
                    spec_path.display(),
                );
                return false
            }
            let error_log = output_directory.join(&api.error_file);
            if error_log.is_file() {
                error!("Dropping API '{}' as it previously failed with generator errors, see '{}' for details.", api.crate_name, error_log.display());
                return false;
            }
            true
        });
        self
    }
    pub fn from_index(index: ApiIndexV1) -> Result<Self, Error> {
        Ok(MappedIndex {
            standard: Standard::default(),
            api: index
                .items
                .into_iter()
                .map(Api::try_from)
                .collect::<Result<Vec<_>, Error>>()?,
        })
    }
}

pub fn crate_name(name: &str, version: &str) -> Result<String, Error> {
    make_target(name, version).map(|n| format!("google-{}", n))
}

/// Currently does the following
/// * strip off all numbers from the tail, until the first non-digit is found
pub fn sanitized_name(name: &str) -> &str {
    if let Some(pos) = name.rfind(|c| !char::is_digit(c, 10)) {
        &name[..=pos]
    } else {
        name
    }
}

pub fn make_target(name: &str, version: &str) -> Result<String, Error> {
    Ok(format!(
        "{name}{version}",
        name = sanitized_name(name),
        version = parse_version(version)?
    ))
}

/// Normalize the version string to adhere to a standard format.
/// The latter could certainly be expressed here at some point, right
/// now it's implied by the tests.
pub fn parse_version(version: &str) -> Result<String, Error> {
    let inner = |version: &str| {
        if version.len() < 2 {
            bail!("version string too small");
        }
        if !version.is_ascii() {
            bail!("can only handle ascii versions");
        }
        if version == "alpha" || version == "beta" {
            return Ok(version.into());
        }

        fn transform_version(version: &str) -> Result<String, Error> {
            let mut bytes = version.bytes();
            if bytes.next() != Some(b'v') {
                bail!("A version must start with 'v'");
            }
            let mut out = String::new();
            let mut separator = Some('_');
            for b in bytes {
                let c = match b {
                    b'.' => b'd',
                    b @ b'0'..=b'9' => b,
                    b @ b'a'..=b'z' => {
                        if let Some(sep) = separator.take() {
                            out.push(sep);
                        }
                        b
                    }
                    b => bail!("unexpected character '{}'", b),
                } as char;
                out.push(c);
            }
            Ok(out)
        }

        let mut tokens = version.splitn(2, '_');
        if let (Some(left), Some(right)) = (tokens.next(), tokens.next()) {
            return Ok(format!(
                "{version}_{name}",
                version = transform_version(right)?,
                name = left
            ));
        }
        transform_version(version)
    };
    inner(version).map_err(|e| format_err!("invalid version '{}': {}", version, e))
}
