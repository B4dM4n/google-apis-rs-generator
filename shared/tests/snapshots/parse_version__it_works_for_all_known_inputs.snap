---
created: "2019-08-19T08:12:56.238951Z"
creator: insta@0.10.1
source: mcp/src/shared.rs
expression: expected
---
alpha alpha
beta beta
datatransfer_v1 1_datatransfer
directory_v1 1_directory
reports_v1 1_reports
v1 1
v1.1 1d1
v1.2 1d2
v1.3 1d3
v1.4 1d4
v1alpha 1_alpha
v1alpha1 1_alpha1
v1alpha2 1_alpha2
v1b3 1_b3
v1beta 1_beta
v1beta1 1_beta1
v1beta1a 1_beta1a
v1beta2 1_beta2
v1beta3 1_beta3
v1beta4 1_beta4
v1beta5 1_beta5
v1configuration 1_configuration
v1management 1_management
v1p1alpha1 1_p1alpha1
v1p1beta1 1_p1beta1
v1p2beta1 1_p2beta1
v1p3beta1 1_p3beta1
v2 2
v2.1 2d1
v2.4 2d4
v2alpha1 2_alpha1
v2beta 2_beta
v2beta1 2_beta1
v2beta2 2_beta2
v2beta3 2_beta3
v3 3
v3.1 3d1
v3.2 3d2
v3.3 3d3
v3beta1 3_beta1
v3p1alpha1 3_p1alpha1
v3p1beta1 3_p1beta1
v4 4
v4.1 4d1
v5 5
